README.md
# moxa-gpio-control

`moxa-gpio-control` is a C library for controlling the Linux GPIO
subsystem via sysfs interface.

## Build

This project use autotools as buildsystem. You can build this project by

* If the build target architecture is x86_64

	```
	# ./autogen.sh --host=x86_64-linux-gnu --includedir=/usr/include/moxa --libdir=/usr/lib/x86_64-linux-gnu
	# make
	# make install
	```
* If the build target architecture is armhf

	```
	# ./autogen.sh --host=arm-linux-gnueabihf --includedir=/usr/include/moxa --libdir=/usr/lib/arm-linux-gnueabihf
	# make
	# make install
	```

The autogen script will execute ./configure and pass all the command-line
arguments to it.

## Documentation

[API Reference](/API_References.md)


debian/changelog
moxa-gpio-control (1.0.3+deb10) buster; urgency=medium

  * Initial release
  * Source code was based on version 1.0.3

 -- Elvis Yao (姚家瑋) <ElvisCW.Yao@moxa.com>  Tue, 21 Jul 2020 19:37:22 +0800


example/gpio-test.c
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Name:
 *	C Example code for MOXA GPIO Library
 *
 * Description:
 *	Example code for demonstrating the usage of GPIO Library in C
 *
 * Authors:
 *	2019	Ken CJ Chou	<KenCJ.Chou@moxa.com>
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <moxa/mx_gpio.h>

static int is_exported_at_first;

static void exit_reset_gpio(int exit_code, int gpio_num)
{
	if (!is_exported_at_first) {
		printf("Unexport GPIO %d.\n", gpio_num);
		mx_gpio_unexport(gpio_num);
	}

	exit(exit_code);
}

int main(int argc, char *argv[])
{

	int gpio_num, ret;
	int direction, value;

	if (argc != 2) {
		fprintf(stderr, "Usage: %s <gpio_num>\n", argv[0]);
		exit(99);
	}
	gpio_num = atoi(argv[1]);

	printf("Testing GPIO %d\n", gpio_num);
	printf("====================\n");

	if (mx_gpio_is_exported(gpio_num)) {
		printf("- GPIO %d is exported.\n", gpio_num);
		is_exported_at_first = 1;
	} else {
		printf("- GPIO %d is not exported,\n", gpio_num);
		printf("  export GPIO %d now.\n", gpio_num);

		ret = mx_gpio_export(gpio_num);
		if (ret < 0) {
			fprintf(stderr, "Error: Failed to export GPIO %d\n", gpio_num);
			fprintf(stderr, "Return code: %d\n", ret);
			exit(1);
		}
		is_exported_at_first = 0;
	}

	ret = mx_gpio_get_direction(gpio_num, &direction);
	if (ret < 0) {
		fprintf(stderr, "Error: Failed to get GPIO %d's direction\n", gpio_num);
		fprintf(stderr, "Return code: %d\n", ret);
		exit_reset_gpio(1, gpio_num);
	}
	ret = mx_gpio_get_value(gpio_num, &value);
	if (ret < 0) {
		fprintf(stderr, "Error: Failed to get GPIO %d's value\n", gpio_num);
		fprintf(stderr, "Return code: %d\n", ret);
		exit_reset_gpio(1, gpio_num);
	}
	printf("  The direction is %d\n", direction);
	printf("  The value is %d\n", value);

	printf("- Setting GPIO %d to \"out\" and \"high\"\n", gpio_num);
	ret = mx_gpio_set_direction(gpio_num, GPIO_DIRECTION_OUT);
	if (ret < 0) {
		fprintf(stderr, "Error: Failed to set GPIO %d's direction\n", gpio_num);
		fprintf(stderr, "Return code: %d\n", ret);
		exit_reset_gpio(1, gpio_num);
	}
	ret = mx_gpio_set_value(gpio_num, GPIO_VALUE_HIGH);
	if (ret < 0) {
		fprintf(stderr, "Error: Failed to set GPIO %d's value\n", gpio_num);
		fprintf(stderr, "Return code: %d\n", ret);
		exit_reset_gpio(1, gpio_num);
	}
	sleep(2);

	ret = mx_gpio_get_direction(gpio_num, &direction);
	if (ret < 0) {
		fprintf(stderr, "Error: Failed to get GPIO %d's direction\n", gpio_num);
		fprintf(stderr, "Return code: %d\n", ret);
		exit_reset_gpio(1, gpio_num);
	}
	ret = mx_gpio_get_value(gpio_num, &value);
	if (ret < 0) {
		fprintf(stderr, "Error: Failed to get GPIO %d's value\n", gpio_num);
		fprintf(stderr, "Return code: %d\n", ret);
		exit_reset_gpio(1, gpio_num);
	}
	printf("  The direction is %d\n", direction);
	printf("  The value is %d\n", value);

	sleep(1);
	printf("====================\n");
	printf("Test OK.\n");
	exit_reset_gpio(0, gpio_num);
}


include/mx_gpio.h
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Name:
 *	MOXA GPIO Library
 *
 * Description:
 *	Library for GPIO.
 *
 * Authors:
 *	2018	Ken CJ Chou	<KenCJ.Chou@moxa.com>
 */

#ifndef _MOXA_GPIO_H
#define _MOXA_GPIO_H

enum gpio_direction {
	GPIO_DIRECTION_IN = 0,
	GPIO_DIRECTION_OUT = 1
};

enum gpio_value {
	GPIO_VALUE_LOW = 0,
	GPIO_VALUE_HIGH = 1
};

#ifdef __cplusplus
extern "C" {
#endif

extern int mx_gpio_is_exported(int gpio_num);
extern int mx_gpio_export(int gpio_num);
extern int mx_gpio_unexport(int gpio_num);
extern int mx_gpio_set_direction(int gpio_num, int direction);
extern int mx_gpio_get_direction(int gpio_num, int *direction);
extern int mx_gpio_set_value(int gpio_num, int value);
extern int mx_gpio_get_value(int gpio_num, int *value);

#ifdef __cplusplus
}
#endif

#endif /* _MOXA_GPIO_H */



lib/mx_gpio.c
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Name:
 *	MOXA GPIO Library
 *
 * Description:
 *	Library for GPIO.
 *
 * Authors:
 *	2018	Ken CJ Chou	<KenCJ.Chou@moxa.com>
 */

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include <mx_gpio.h>

#define SYSFS_GPIO_PATH "/sys/class/gpio"
#define MAX_FILEPATH_LEN 256	/* reserved length for file path */
#define MAX_BUFFER_LEN 32	/* reserved length for buffer */

/*
 * static functions
 */
static int is_gpio_exported(int gpio_num)
{
	char filepath[MAX_FILEPATH_LEN];

	/* check if value file existed */
	sprintf(filepath, "%s/gpio%d/value", SYSFS_GPIO_PATH, gpio_num);
	if (access(filepath, F_OK) < 0)
		return 0;

	return 1;
}

static int read_file(char *filepath, char *data)
{
	int fd;

	fd = open(filepath, O_RDONLY);
	if (fd < 0) 
		return -1; /* E_SYSFUNCERR */

	if (read(fd, data, sizeof(data)) <= 0) {
		close(fd);
		return -1; /* E_SYSFUNCERR */
	}
	close(fd);
	return 0;
}

static int write_file(char *filepath, const char *data)
{
	int fd;

	fd = open(filepath, O_WRONLY);
	if (fd < 0)
		return -1; /* E_SYSFUNCERR */

	if (write(fd, data, strlen(data)) < 0) {
		close(fd);
		return -1; /* E_SYSFUNCERR */
	}
	close(fd);
	return 0;
}

/*
 * APIs
 */
int mx_gpio_is_exported(int gpio_num)
{
	return is_gpio_exported(gpio_num);
}

int mx_gpio_export(int gpio_num)
{
	char filepath[MAX_FILEPATH_LEN];
	char buffer[MAX_BUFFER_LEN];

	/* if device node is existed, do nothing */
	if (is_gpio_exported(gpio_num))
		return 0;

	sprintf(filepath, "%s/export", SYSFS_GPIO_PATH);
	sprintf(buffer, "%d", gpio_num);

	return write_file(filepath, buffer);
}

int mx_gpio_unexport(int gpio_num)
{
	char filepath[MAX_FILEPATH_LEN];
	char buffer[MAX_BUFFER_LEN];

	/* if device node is not existed, do nothing */
	if (!is_gpio_exported(gpio_num))
		return 0;

	sprintf(filepath, "%s/unexport", SYSFS_GPIO_PATH);
	sprintf(buffer, "%d", gpio_num);

	return write_file(filepath, buffer);
}

int mx_gpio_set_direction(int gpio_num, int direction)
{
	char filepath[MAX_FILEPATH_LEN];

	if (mx_gpio_export(gpio_num))
		return -1; /* E_SYSFUNCERR */

	if (!is_gpio_exported(gpio_num))
		return -20; /* E_GPIO_NOTEXP */

	sprintf(filepath, "%s/gpio%d/direction", SYSFS_GPIO_PATH, gpio_num);
	if (direction == GPIO_DIRECTION_IN) {
		return write_file(filepath, "in");
	} else if (direction == GPIO_DIRECTION_OUT) {
		return write_file(filepath, "out");
	} else {
		return -2; /* E_INVAL */
	}
}

int mx_gpio_get_direction(int gpio_num, int *direction)
{
	char filepath[MAX_FILEPATH_LEN];
	char buffer[MAX_BUFFER_LEN];
	int ret;

	if (mx_gpio_export(gpio_num))
		return -1; /* E_SYSFUNCERR */

	if (!is_gpio_exported(gpio_num))
		return -20; /* E_GPIO_NOTEXP */

	sprintf(filepath, "%s/gpio%d/direction", SYSFS_GPIO_PATH, gpio_num);

	ret = read_file(filepath, buffer);
	if (ret < 0)
		return ret;

	if (!strncmp(buffer, "in", 2)) {
		*direction = GPIO_DIRECTION_IN;
		return 0;
	} else if (!strncmp(buffer, "out", 3)) {
		*direction = GPIO_DIRECTION_OUT;
		return 0;
	}
	return -21; /* E_GPIO_UNKDIR */
}

int mx_gpio_set_value(int gpio_num, int value)
{
	char filepath[MAX_FILEPATH_LEN];

	if (mx_gpio_export(gpio_num))
		return -1; /* E_SYSFUNCERR */

	if (!is_gpio_exported(gpio_num))
		return -20; /* E_GPIO_NOTEXP */

	sprintf(filepath, "%s/gpio%d/direction", SYSFS_GPIO_PATH, gpio_num);
	if (value == GPIO_VALUE_LOW) {
		return write_file(filepath, "low");
	} else if (value == GPIO_VALUE_HIGH) {
		return write_file(filepath, "high");
	} else {
		return -2; /* E_INVAL */
	}
}

int mx_gpio_get_value(int gpio_num, int *value)
{
	char filepath[MAX_FILEPATH_LEN];
	char buffer[MAX_BUFFER_LEN];
	int ret;

	if (mx_gpio_export(gpio_num))
		return -1; /* E_SYSFUNCERR */

	if (!is_gpio_exported(gpio_num))
		return -20; /* E_GPIO_NOTEXP */

	sprintf(filepath, "%s/gpio%d/value", SYSFS_GPIO_PATH, gpio_num);

	ret = read_file(filepath, buffer);
	if (ret < 0)
		return ret;

	if (!strncmp(buffer, "0", 1)) {
		*value = GPIO_VALUE_LOW;
		return 0;
	} else if (!strncmp(buffer, "1", 1)) {
		*value = GPIO_VALUE_HIGH;
		return 0;
	}
	return -22; /* E_GPIO_UNKVAL */
}
